homeassistant/components/abode/config_flow.py

```diff
@@ -26,48 +27,151 @@ def __init__(self):
             vol.Required(CONF_PASSWORD): str,
         }
 
+        self.mfa_data_schema = {
+            vol.Required(CONF_MFA): str,
+        }
+
+        self._username = None
+        self._password = None
+        self._polling = None
+        self._cache = None
+
+    async def _async_abode_login(self, step_id):
+        """Handle login with Abode."""
+        self._cache = self.hass.config.path(DEFAULT_CACHEDB)
+        errors = {}
+
+        try:
+            abode = Abode(
+                auto_login=False,
+                get_devices=False,
+                get_automations=False,
+                cache_path=self._cache,
+            )
+            await self.hass.async_add_executor_job(
+                abode.login, self._username, self._password
+            )
+
+        except (AbodeException, ConnectTimeout, HTTPError) as ex:
+            if ex.errcode == MFA_CODE_REQUIRED[0]:
+                return await self.async_step_mfa()
+
+            LOGGER.error("Unable to connect to Abode: %s", ex)
+
+            if ex.errcode == HTTP_BAD_REQUEST:
+                errors = {"base": "invalid_auth"}
+            else:
+                errors = {"base": "cannot_connect"}
+
+        if errors:
+            return self.async_show_form(
+                step_id=step_id, data_schema=vol.Schema(self.data_schema), errors=errors
+            )
+
+        return await self._async_create_entry(
+            {
+                CONF_USERNAME: self._username,
+                CONF_PASSWORD: self._password,
+                CONF_POLLING: self._polling,
+            }
+        )
+
+    async def _async_create_entry(self, user_input):
```

Comment:
This was previously `async_finish_flow` but I instead changed it to a helper function as well since it's just handling the creation of the config entry.